[disk] AgentCheck for Disk metrics :tada:

degemer opened the pull request:
This commit is the first step towards the annihilation of the old `Check`
hidden in `checks/system/{unix,win32}.py`!
It's time to launch the final assault of `Check`. `AgentCheck` shall triumph!

Add a `checks.d/disk.py` checks, which uses psutil when available to get
disk metrics (same than the previous `disk` checks), and otherwise
get the same metrics with `df`.
It has a small integration test, just to check that metrics exist, and
more precise mock tests.
This new check is enabled by default, and the old `disk` checks are
deleted.

Fix https://github.com/DataDog/dd-agent/issues/568

remh commented on checks.d/disk.py:
```diff
@@ -0,0 +1,229 @@
+        list devices. A 'device' is a list with fields corresponding to the
+        output of df output on each platform.
+        """
+        all_devices = [l.strip().split() for l in df_output.split("\n")]
```
I know it doesn't matter in this case as it's only for linux but you should use:
```python
df_output.splitlines()
```
instead. It's better practice as it should work on all platforms.

LeoCavaille commented on checks.d/disk.py:
```diff
@@ -0,0 +1,229 @@
+        else:
+            self.collect_metrics_manually()
+
+    def _psutil(self):
```
Should be a static method or a class method.

LeoCavaille commented on checks.d/disk.py:
```diff
@@ -0,0 +1,229 @@
+        self._tag_by_filesystem = _is_affirmative(
+            instance.get('tag_by_filesystem', ''))
+        self._all_partitions = _is_affirmative(
+            instance.get('all_partitions', 'yes'))
```
Just for clarity here I'd recommend using a boolean directly rather than a string for the default.

LeoCavaille commented on checks.d/disk.py:
```diff
@@ -0,0 +1,229 @@
+        legacy_option = self.agentConfig.get(legacy_name)
+        if self_option == default and legacy_option is not None:
+            self_option = legacy_option
+        setattr(self, '_{0}'.format(option), operation(self_option))
```
```python

LeoCavaille replied:
Could be factorized a bit and also throw a deprecation warning?
```python
value = instance.get(option)
legacy_name = legacy_name or option
if value is None and legacy_name in self.agentConfig:
    self.log.warn(
        "Using `{0}` in datadog.conf has been deprecated in favor of `{1}` in disk.yaml"
        .format(legacy_name, option)
    )
    value = self.agentConfig(legacy_name)
setattr(self, '_{0}'.format(option), operation(value))
```

remh replied:
(watchout not to spam the log with this deprecation warning )

LeoCavaille commented on checks.d/disk.py:
```diff
@@ -0,0 +1,229 @@
+            # we check all exclude conditions
+            if self._exclude_disk_psutil(part):
+                continue
+            else:
```
Lose the `else:` just continue the loop for simplicity.

LeoCavaille commented on checks.d/disk.py:
```diff
@@ -0,0 +1,229 @@
+                continue
+            else:
+                # For later, latency metrics
+                self._valid_disks[part.device] = [part.fstype, part.mountpoint]
```
Use a tuple and not a list for a fixed-size structure.

LeoCavaille commented on checks.d/disk.py:
```diff
@@ -0,0 +1,229 @@
+        if Platform.is_unix():
+            metrics.update(self._collect_inodes_metrics(part.mountpoint))
+
+        return metrics.iteritems()
```
I am not sure it's adding much here to return a generator, you should return a dictionary and iterate on it in the caller.
Otherwise it's confusing some `_collect_*` methods return dicts, other return dictionaries or call `self.gauge` directly.

degemer commented:
I addressed all your comments, moved `_load_conf` to `__init__` (it would be nice to  put everything in `init_config`, but we wouldn't have `instances`, and I guess the check would not run so...).
Also:
* `collect_*` actually send metrics (`self.gauge`),
* `_collect_*` returns dictionary of metrics -> value.

LeoCavaille commented:
Thanks for making the updates. Merging.
